fl4health/clients/ditto_mkmmd_client.py

```diff
@@ -0,0 +1,483 @@
+                from server config, also `mkmmd_loss`, `feature_l2_norm_loss` keys and their respective calculated
+                values.
+        """
+
+        # Compute global model vanilla loss
+        assert "global" in preds
+        global_loss = self.criterion(preds["global"], target)
+
+        # Compute local model loss + ditto constraint term
+        assert "local" in preds
+        local_loss = self.criterion(preds["local"], target)
+        total_loss = local_loss.clone()
+
+        additional_losses = {"local_loss": local_loss, "global_loss": global_loss}
+
+        if self.mkmmd_loss_weight != 0:
+            assert "init_global_features" in features
+            assert "features" in features
+            # Compute MK-MMD loss
+            mkmmd_loss = self.mkmmd_loss(features["features"], features["init_global_features"])
+            total_loss += self.mkmmd_loss_weight * mkmmd_loss
+            additional_losses["mkmmd_loss"] = mkmmd_loss
+            if self.feature_l2_norm:
+                feature_l2_norm_loss = torch.linalg.norm(features["features"])
+                total_loss += self.feature_l2_norm * feature_l2_norm_loss
+                additional_losses["feature_l2_norm_loss"] = feature_l2_norm_loss
+
+        additional_losses["total_loss"] = total_loss
+
+        return total_loss, additional_losses
+
+    def compute_training_loss(
+        self,
+        preds: Dict[str, torch.Tensor],
+        features: Dict[str, torch.Tensor],
+        target: torch.Tensor,
+    ) -> TrainingLosses:
+        """
+        Computes training losses given predictions of the global and local models and ground truth data.
+        For the local model we add to vanilla loss function by including Ditto penalty loss which is the l2 inner
+        product between the initial global model weights and weights of the local model. This is stored in backward
+        The loss to optimize the global model is stored in the additional losses dictionary under "global_loss"
+
+        Args:
+            preds (Dict[str, torch.Tensor]): Prediction(s) of the model(s) indexed by name.
+                All predictions included in dictionary will be used to compute metrics.
+            features: (Dict[str, torch.Tensor]): Feature(s) of the model(s) indexed by name.
+            target: (torch.Tensor): Ground truth data to evaluate predictions against.
+
+        Returns:
+            TrainingLosses: an instance of TrainingLosses containing backward loss and
+                additional losses indexed by name. Additional losses includes each loss component and the global model
+                loss tensor.
+        """
+        # Check that both models are in training mode
+        assert self.global_model.training and self.model.training
+
+        total_loss, additional_losses = self.compute_loss_and_additional_losses(preds, features, target)
```

Comment:
If we refactor the Ditto class to use this method and override `self.compute_loss_and_additional_losses` to compute the global and local losses there, we can just drop this from this client after inheriting from Ditto to save code duplication.